File: Arrays 101/Squares of a Sorted Array.java
class Solution {
	public int[] sortedSquares(int[] A) {

		int right = A.length - 1;
		int left = 0;

		int[] result = new int[A.length];
		int resultIndex = result.length - 1;

    	while (left <= right) { // !!! <=, < will result in one element being skipped. 
    		int leftSquare = A[left] * A[left];
    		int rightSquare = A[right] * A[right];

    		if (leftSquare > rightSquare) {
    			result[resultIndex] = leftSquare;
    			left ++;
    		} else {
    			result[resultIndex] = rightSquare;
    			right --;
    		}
    		resultIndex --;
    	}
    	return result;

    }

    // Solution
    public int[] sortedSquares(int[] A) {
    	int N = A.length;
    	int j = 0;
    	while (j < N && A[j] < 0) {
    		j ++;
    	}
    	int i = j - 1;
    	// i is the index of the last negative number
    	// j is the index of the first positive number

    	int[] ans = new int[N];
    	int t = 0;

    	while (i >= 0 && j < N) {
    		if (A[i] * A[i] < A[j] * A[j]) {
    			ans[t++] = A[i] * A[i];
    			i --;
    		} else {
    			ans[t++] = A[j] * A[j];
    			j ++;
    		}
    	}

    	// fill in the leftovers from either side
    	while (i >= 0) {
    		ans[t++] = A[i] * A[i];
    	}

    	while (j < N) {
    		ans[t++] = A[j] * A[j];

    		
    	}
    	return ans;
    }

}

File: Two Pointer/Squares of a Sorted Array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:

        result = []

        left = 0
        right = len(nums) - 1

        # <=, try writing out an example
        # if <, then the = value will be skipped
        while left <= right:

            left_square = nums[left] * nums[left]
            right_square = nums[right] * nums[right]

            if left_square > right_square:

                result.append(left_square)
                left += 1

            else:

                result.append(right_square)
                right -= 1

        # we built up solution from greatest to least, because
        # the "outsides" of the sorted list indicate the greatest values
        result.reverse()
        
        return result





